python/src/opendp/typing.py

```diff
@@ -268,13 +269,15 @@ def infer(cls, public_example: Any) -> Union["RuntimeType", str]:
             return RuntimeType.parse(public_example.type)
 
         if isinstance(public_example, tuple):
-            return RuntimeType('Tuple', list(map(cls.infer, public_example)))
+            return RuntimeType('Tuple', [cls.infer(e, py_object) for e in public_example])
 
         if isinstance(public_example, list):
             if public_example:
-                inner_type = cls.infer(public_example[0])
-                for inner in public_example[1:]:
-                    other_type = cls.infer(inner)
+                inner_type = cls.infer(public_example[0], py_object=py_object)
+                for inner in public_example:
+                    other_type = cls.infer(inner, py_object)
+                    if inner_type == "ExtrinsicObject":
+                        break
```

Comment:
Issue here: We want to be able to detect inadvertent heterogeneous list, but also allow intentional heterogeneous list in the case of ExtrinsicObject.

From discussion:
 * Change `infer` to return Vec<ExtrinsicObject> if heterogeneous and py_object==True; no error message here in that case.
 * Change `assert_is_similar` to take example, do the right thing in case of heterogeneous list.

Reply:
I got rid of assert_is_similar completely, instead folding it directly into the loaders.